---
layout: post
category : Metagenomics, pipelines, publications
tagline: ""
tags : [Shotgun metagenomics, bioinformatics, high performance computing, metagenomics, Shallow sequencing]
---
{% include JB/setup %}

## Shotgun Metagenomics: Determining Depth 
How much sequencing depth is needed for shotgun metagenomic sequencing projects has pretty much always been a controversial topic. We just published an article in [Briefings in Bioinformatics](https://doi.org/10.1093/bib/bbac443) that provides really interesting results regarding the use of shallow sequencing for shotgun metagenomics. Briefly, we demonstrate that shallow sequencing yields more than enough data for three projects/datasets considered in our study. Hitting the sweetspot in sequencing output per sample depends on the complexity of the resident microbial communities and the total number of samples. For instance, for a 912 samples human gut microbiome project initially having 12 terabases of raw sequencing data (which corresponds to an average of ~13 Gigabases/sample), we showed that only ~1 gigabase per sample was needed to provide microbial ecology metrics similar to what was given using the whole sequencing dataset.

